File: cloud/router/src/routes/messages.js
// POST /v1/messages — Anthropic-compatible entry point.
// Reuses the entire chat-completions pipeline (auth, rate limit, combo,
// account picking, cooldown, usage). The actual upstream call is selected
// by provider:
//   * provider === 'claude'   → claude.ai subscription executor (Anthropic
//     native passthrough with spoof headers).
//   * everything else         → OpenAI-compatible client; the Anthropic body
//     is translated to OpenAI and the response back to Anthropic.

import crypto from 'node:crypto';
import {
  logger, ValidationError, UpstreamError, AuthError,
  NoAccountAvailableError, AppError,
} from '@lazirouter-cloud/shared';
import { enforceDailyQuota } from '@lazirouter-cloud/shared';
import { resolveApiKey, enforceTenantActive } from '../middleware/edgeAuth.js';
import { enforceRateLimit } from '../middleware/rateLimit.js';
import { pickAccount, markCooldown } from '../services/accountPicker.js';
import { recordUsage, getPricing, computeCost } from '../services/usage.js';
import { resolveAttempts } from '../services/combo.js';
import { classifyScenario } from '../services/scenarioClassifier.js';
import { resolveScenarioTarget } from '../services/scenarioRouter.js';
import { getUpstream, isAnthropicNative } from '../providers/index.js';
import {
  anthropicToOpenaiRequest,
  openaiToAnthropicResponse,
  createOpenaiToAnthropicStream,
} from '../translator/anthropic.js';

export async function handleMessages(req, res) {
  const startedAt = Date.now();
  const requestId = crypto.randomUUID();

  const authHeader = req.headers['authorization'] || '';
  const apiKey = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : '';
  // Anthropic SDKs use the `x-api-key` header instead of Authorization
  const xApi = req.headers['x-api-key'];
  const effectiveKey = apiKey || (typeof xApi === 'string' ? xApi : '');
  if (!effectiveKey) throw new AuthError('Missing Authorization or x-api-key header');
  const ctx = await resolveApiKey(effectiveKey);
  enforceTenantActive(ctx);
  await enforceRateLimit(ctx);
  await enforceDailyQuota(ctx);

  const anthropicBody = await readJson(req);
  if (!Array.isArray(anthropicBody.messages)) throw new ValidationError('Missing field: messages[]');

  // Auto-routing: 'auto' or empty model triggers content-based scenario routing.
  const rawModel = (anthropicBody.model ?? '').toString();
  let effectiveModel = rawModel;
  let routedScenario = null;
  if (rawModel === '' || rawModel === 'auto') {
    routedScenario = classifyScenario(anthropicBody, 'anthropic');
    effectiveModel = await resolveScenarioTarget(ctx.tenantId, routedScenario);
    logger.info({ tenantId: ctx.tenantId, scenario: routedScenario, effectiveModel }, 'auto-routing');
  }

  const attempts = await resolveAttempts(ctx.tenantId, effectiveModel);
  const isStream = anthropicBody.stream === true;

  const errors = [];
  for (let i = 0; i < attempts.length; i++) {
    const attempt = attempts[i];
    let account;
    try {
      account = await pickAccount(ctx.tenantId, attempt.provider);
    } catch (e) {
      if (e instanceof NoAccountAvailableError) {
        errors.push({ step: attempt.step, provider: attempt.provider, reason: e.message });
        continue;
      }
      throw e;
    }

    const upstream = getUpstream(attempt.provider);
    const usesAnthropicShape = isAnthropicNative(attempt.provider);

    // For OpenAI-compat path, require base_url + api_key credentials.
    // For Anthropic-native (claude) path, require access_token from OAuth.
    let baseUrl = null;
    if (upstream.kind === 'openai') {
      baseUrl = account.metadata.base_url || defaultBaseUrl(attempt.provider);
      if (!baseUrl) { errors.push({ step: attempt.step, reason: `no base_url for ${attempt.provider}` }); continue; }
      if (!account.credentials.api_key) { errors.push({ step: attempt.step, reason: 'missing api_key' }); continue; }
    } else if (!account.credentials.access_token) {
      errors.push({ step: attempt.step, reason: 'missing OAuth access_token' });
      continue;
    }

    const upstreamBody = usesAnthropicShape
      ? { ...anthropicBody, model: attempt.upstreamModel }
      : { ...anthropicToOpenaiRequest(anthropicBody), model: attempt.upstreamModel };

    const usageBase = {
      tenantId: ctx.tenantId,
      apiKeyId: ctx.apiKeyId,
      connectionId: account.connectionId,
      provider: attempt.provider,
      model: rawModel || 'auto',
      routedModel: routedScenario ? effectiveModel : null,
      upstreamModel: attempt.upstreamModel,
      requestId,
    };

    try {
      if (isStream) {
        await handleStream({
          req, res, upstream, credentials: account.credentials, metadata: account.metadata,
          baseUrl, upstreamBody,
          usesAnthropicShape, requestedModel: anthropicBody.model,
          usageBase, startedAt,
        });
      } else {
        await handleNonStream({
          res, upstream, credentials: account.credentials, metadata: account.metadata,
          baseUrl, upstreamBody,
          usesAnthropicShape, requestedModel: anthropicBody.model,
          usageBase, startedAt,
        });
      }
      return;
    } catch (err) {
      const status = err instanceof UpstreamError ? err.meta?.status_code : null;
      const retriable = status === 429 || (status >= 500 && status < 600) || !(err instanceof AppError);
      if (retriable && status) await markCooldown(ctx.tenantId, attempt.provider, account.connectionId);
      await recordUsage({
        ...usageBase, status: 'error',
        latencyMs: Date.now() - startedAt,
        errorCode: status ? `upstream_${status}` : (err.code || 'unknown'),
        meta: { error: err.message, attempt: attempt.step, transport: '/v1/messages' },
      });
      errors.push({ step: attempt.step, provider: attempt.provider, status, message: err.message });
      if (res.headersSent) throw err;
      if (!retriable) throw err;
    }
  }
  const allNoAccount = errors.length > 0 && errors.every(e => !e.status);
  if (allNoAccount) {
    throw new NoAccountAvailableError({ attempts: errors });
  }
  throw new UpstreamError('All upstream attempts failed', { attempts: errors });
}

async function handleNonStream({ res, upstream, credentials, metadata, baseUrl, upstreamBody, usesAnthropicShape, requestedModel, usageBase, startedAt }) {
  const result = await upstream.chat({ credentials, metadata, baseUrl, body: upstreamBody });
  const anthropicResp = usesAnthropicShape ? result : openaiToAnthropicResponse(result, requestedModel);

  const usage = result.usage || {};
  const promptTokens = usage.prompt_tokens ?? usage.input_tokens ?? 0;
  const completionTokens = usage.completion_tokens ?? usage.output_tokens ?? 0;
  const pricing = await getPricing(usageBase.tenantId, usageBase.provider, usageBase.upstreamModel);
  const costMicros = computeCost(pricing, promptTokens, completionTokens);

  res.writeHead(200, { 'content-type': 'application/json' });
  res.end(JSON.stringify(anthropicResp));

  await recordUsage({
    ...usageBase, status: 'ok', latencyMs: Date.now() - startedAt,
    promptTokens, completionTokens, costMicros,
    meta: { transport: '/v1/messages' },
  });
}

async function handleStream({ req, res, upstream, credentials, metadata, baseUrl, upstreamBody, usesAnthropicShape, requestedModel, usageBase, startedAt }) {
  const ac = new AbortController();
  req.on('close', () => ac.abort());

  let headersSent = false;
  let promptTokens = 0;
  let completionTokens = 0;
  const xlate = usesAnthropicShape ? null : createOpenaiToAnthropicStream(requestedModel);

  for await (const frame of upstream.stream({ credentials, metadata, baseUrl, body: upstreamBody, signal: ac.signal })) {
    if (!headersSent) {
      res.writeHead(200, {
        'content-type': 'text/event-stream',
        'cache-control': 'no-cache, no-transform',
        'connection': 'keep-alive',
      });
      headersSent = true;
    }
    if (xlate) {
      for (const out of xlate.feed(frame)) res.write(out);
    } else {
      res.write(frame);
    }
    // Sniff usage even for passthrough. Anthropic emits both input_tokens
    // and output_tokens; they can appear in either order across frames so
    // capture them independently and keep the latest seen.
    const pin  = /"(?:prompt_tokens|input_tokens)"\s*:\s*(\d+)/.exec(frame);
    const pout = /"(?:completion_tokens|output_tokens)"\s*:\s*(\d+)/.exec(frame);
    if (pin)  promptTokens = Number(pin[1]);
    if (pout) completionTokens = Number(pout[1]);
  }
  if (xlate) {
    for (const out of xlate.flush()) res.write(out);
    const u = xlate.getUsage();
    if (u.input_tokens)  promptTokens     = u.input_tokens;
    if (u.output_tokens) completionTokens = u.output_tokens;
  }
  res.end();

  const pricing = await getPricing(usageBase.tenantId, usageBase.provider, usageBase.upstreamModel);
  const costMicros = computeCost(pricing, promptTokens, completionTokens);
  await recordUsage({
    ...usageBase, status: 'ok', latencyMs: Date.now() - startedAt,
    promptTokens, completionTokens, costMicros,
    meta: { transport: '/v1/messages' },
  });
}

function defaultBaseUrl(provider) {
  switch (provider) {
    case 'openai':    return 'https://api.openai.com';
    case 'glm':       return 'https://open.bigmodel.cn/api/paas/v4';
    case 'deepseek':  return 'https://api.deepseek.com';
    case 'minimax':   return 'https://api.minimaxi.com';
    case 'anthropic': return 'https://api.anthropic.com';
    default: return null;
  }
}

function readJson(req) {
  return new Promise((resolve, reject) => {
    let buf = '';
    req.on('data', c => { buf += c; if (buf.length > 5 * 1024 * 1024) { req.destroy(); reject(new ValidationError('Body too large')); } });
    req.on('end', () => {
      if (!buf) return resolve({});
      try { resolve(JSON.parse(buf)); } catch { reject(new ValidationError('Invalid JSON body')); }
    });
    req.on('error', reject);
  });
}

File: internal/router/messages.go
// /v1/messages — Anthropic-shape entry point.
// Provider routing:
//   provider == "claude"  → ClaudeSub (Anthropic-native, no body translation)
//   otherwise             → OpenAI-compat upstream + translator (Anthropic↔OpenAI)
package router

import (
	"context"
	"errors"
	"net/http"
	"time"

	"github.com/amaozhao/lazirouter/internal/combo"
	"github.com/amaozhao/lazirouter/internal/errs"
	"github.com/amaozhao/lazirouter/internal/httpx"
	"github.com/amaozhao/lazirouter/internal/logger"
	"github.com/amaozhao/lazirouter/internal/picker"
	"github.com/amaozhao/lazirouter/internal/quota"
	"github.com/amaozhao/lazirouter/internal/scenario"
	"github.com/amaozhao/lazirouter/internal/tenantctx"
	"github.com/amaozhao/lazirouter/internal/translator"
	"github.com/amaozhao/lazirouter/internal/usage"
	"github.com/google/uuid"
)

func (d *Deps) handleMessages(w http.ResponseWriter, r *http.Request) {
	started := time.Now()
	reqID := uuid.NewString()

	apiKey := httpx.BearerOrXAPIKey(r)
	if apiKey == "" {
		httpx.WriteError(w, errs.Auth("Missing Authorization or x-api-key header"))
		return
	}
	ctx, err := tenantctx.ResolveAPIKey(r.Context(), apiKey)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	if err := tenantctx.EnforceTenantActive(ctx); err != nil {
		httpx.WriteError(w, err)
		return
	}
	if err := quota.Enforce(r.Context(), ctx); err != nil {
		httpx.WriteError(w, err)
		return
	}

	body, err := httpx.ReadJSONMap(r)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	if _, ok := body["messages"].([]any); !ok {
		httpx.WriteError(w, errs.Validation("Missing field: messages[]"))
		return
	}
	requestedModel, _ := body["model"].(string)
	model := requestedModel
	if model == "" {
		model = "auto"
	}
	stream := truthy(body["stream"])

	effectiveModel := model
	routedScenario := ""
	if model == "auto" {
		routedScenario = scenario.Classify(body)
		target, err := scenario.ResolveTarget(r.Context(), ctx.TenantID, routedScenario)
		if err != nil {
			httpx.WriteError(w, err)
			return
		}
		effectiveModel = target
		logger.Info("auto-routing", "tenantId", ctx.TenantID, "scenario", routedScenario, "effectiveModel", effectiveModel)
	}

	attempts, err := combo.Resolve(r.Context(), ctx.TenantID, effectiveModel)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}

	var errsList []map[string]any
	for _, attempt := range attempts {
		account, err := picker.Pick(r.Context(), d.Cfg.MasterKey, ctx.TenantID, attempt.Provider)
		if err != nil {
			var ae *errs.AppError
			if errors.As(err, &ae) && ae.Code == "no_account_available" {
				errsList = append(errsList, map[string]any{"step": attempt.Step, "provider": attempt.Provider, "reason": ae.Message})
				continue
			}
			httpx.WriteError(w, err)
			return
		}

		ueBase := &usage.Event{
			TenantID: ctx.TenantID, APIKeyID: ctx.APIKeyID, ConnectionID: account.ConnectionID,
			Provider: attempt.Provider, Model: model, UpstreamModel: attempt.UpstreamModel,
			RequestID: reqID, Meta: map[string]any{"transport": "/v1/messages"},
		}
		if routedScenario != "" {
			ueBase.RoutedModel = &effectiveModel
		}

		var callErr error
		switch {
		case attempt.Provider == "claude":
			// Anthropic-native pass-through via Claude subscription executor.
			body["model"] = attempt.UpstreamModel
			if stream {
				callErr = d.streamClaude(r.Context(), w, account, body, ueBase, started, ctx.TenantID, account.ConnectionID)
			} else {
				callErr = d.nonStreamClaude(r.Context(), w, account, body, ueBase, started, ctx.TenantID, requestedModel)
			}
		case stringFrom(account.Metadata, "protocol") == "anthropic_passthrough":
			// Third-party Anthropic-compat endpoint (e.g. MiniMax Token Plan).
			// No translation; forward Anthropic body verbatim, auth via x-api-key.
			baseURL := stringFrom(account.Metadata, "base_url")
			upstreamKey := stringFrom(account.Credentials, "api_key")
			if baseURL == "" || upstreamKey == "" {
				errsList = append(errsList, map[string]any{"step": attempt.Step, "reason": "anthropic_passthrough requires base_url + api_key"})
				continue
			}
			body["model"] = attempt.UpstreamModel
			if stream {
				callErr = d.streamMessagesViaAnthropicCompat(r.Context(), w, baseURL, upstreamKey, account.Metadata, body, ueBase, started)
			} else {
				callErr = d.nonStreamMessagesViaAnthropicCompat(r.Context(), w, baseURL, upstreamKey, account.Metadata, body, ueBase, started)
			}
		default:
			// OpenAI-compatible upstream via translator.
			baseURL := stringFrom(account.Metadata, "base_url")
			if baseURL == "" {
				baseURL = httpx.DefaultBaseURL(attempt.Provider)
			}
			upstreamKey := stringFrom(account.Credentials, "api_key")
			if baseURL == "" || upstreamKey == "" {
				errsList = append(errsList, map[string]any{"step": attempt.Step, "reason": "connection missing base_url or api_key"})
				continue
			}
			openaiBody := translator.AnthropicToOpenAIRequest(body)
			openaiBody["model"] = attempt.UpstreamModel
			if stream {
				callErr = d.streamMessagesViaOpenAI(r.Context(), w, baseURL, upstreamKey, openaiBody, requestedModel, ueBase, started)
			} else {
				callErr = d.nonStreamMessagesViaOpenAI(r.Context(), w, baseURL, upstreamKey, openaiBody, requestedModel, ueBase, started)
			}
		}

		if callErr == nil {
			return
		}

		var ae *errs.AppError
		status := 0
		if errors.As(callErr, &ae) {
			if sc, ok := ae.Meta["status_code"].(int); ok {
				status = sc
			}
		}
		retriable := status == 429 || (status >= 500 && status < 600) || !errors.As(callErr, &ae)
		if retriable && status > 0 {
			_, _ = picker.MarkCooldown(r.Context(), ctx.TenantID, attempt.Provider, account.ConnectionID)
		}
		errsList = append(errsList, map[string]any{"step": attempt.Step, "provider": attempt.Provider, "status": status, "message": callErr.Error()})
		if !retriable {
			httpx.WriteError(w, callErr)
			return
		}
	}

	httpx.WriteError(w, errs.Upstream("All upstream attempts failed", map[string]any{"attempts": errsList}))
}

// --- claude subscription paths ---

func (d *Deps) nonStreamClaude(ctx context.Context, w http.ResponseWriter, account *picker.Account, body map[string]any, ue *usage.Event, started time.Time, tenantID int64, requestedModel string) error {
	result, err := d.ClaudeSub.ChatMessages(ctx, account.Credentials, account.Metadata, body)
	if err != nil {
		_ = usage.Record(ctx, errEvent(ue, started, err))
		return err
	}
	u, _ := result["usage"].(map[string]any)
	prompt := toI64Or(u, "input_tokens")
	completion := toI64Or(u, "output_tokens")
	pricing := usage.GetPricing(ctx, tenantID, "claude", ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	httpx.WriteJSON(w, http.StatusOK, result)
	_ = requestedModel
	return nil
}

func (d *Deps) streamClaude(ctx context.Context, w http.ResponseWriter, account *picker.Account, body map[string]any, ue *usage.Event, started time.Time, tenantID int64, connID int64) error {
	frames, errc := d.ClaudeSub.StreamMessages(ctx, account.Credentials, account.Metadata, body)
	var prompt, completion int64
	headersSent := false
	for {
		select {
		case frame, ok := <-frames:
			if !ok {
				goto done
			}
			if !headersSent {
				w.Header().Set("content-type", "text/event-stream")
				w.Header().Set("cache-control", "no-cache, no-transform")
				w.Header().Set("connection", "keep-alive")
				w.WriteHeader(http.StatusOK)
				headersSent = true
			}
			_, _ = w.Write(frame)
			if f, ok := w.(http.Flusher); ok {
				f.Flush()
			}
			// Claude SSE has `event: message_delta\ndata: {"usage":{...}}`
			p, c := scanAnthropicUsage(frame)
			if p > 0 {
				prompt = p
			}
			if c > 0 {
				completion = c
			}
		case err := <-errc:
			if !headersSent {
				_ = usage.Record(ctx, errEvent(ue, started, err))
				return err
			}
			logger.Warn("claude stream error after headers sent", "err", err)
			goto done
		case <-ctx.Done():
			return nil
		}
	}
done:
	pricing := usage.GetPricing(ctx, tenantID, "claude", ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	_ = connID
	return nil
}

// --- openai-compat paths (translate request and response) ---

func (d *Deps) nonStreamMessagesViaOpenAI(ctx context.Context, w http.ResponseWriter, baseURL, apiKey string, openaiBody map[string]any, requestedModel string, ue *usage.Event, started time.Time) error {
	result, err := d.OpenAI.ChatCompletion(ctx, baseURL, apiKey, openaiBody)
	if err != nil {
		_ = usage.Record(ctx, errEvent(ue, started, err))
		return err
	}
	anthropicResp := translator.OpenAIToAnthropicResponse(result, requestedModel)
	u, _ := result["usage"].(map[string]any)
	prompt := toI64Or(u, "prompt_tokens")
	completion := toI64Or(u, "completion_tokens")
	pricing := usage.GetPricing(ctx, ue.TenantID, ue.Provider, ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	httpx.WriteJSON(w, http.StatusOK, anthropicResp)
	return nil
}

func (d *Deps) streamMessagesViaOpenAI(ctx context.Context, w http.ResponseWriter, baseURL, apiKey string, openaiBody map[string]any, requestedModel string, ue *usage.Event, started time.Time) error {
	frames, errc := d.OpenAI.StreamChatCompletion(ctx, baseURL, apiKey, openaiBody)
	xlate := translator.NewStream(requestedModel)
	headersSent := false
	for {
		select {
		case frame, ok := <-frames:
			if !ok {
				goto flush
			}
			if !headersSent {
				w.Header().Set("content-type", "text/event-stream")
				w.Header().Set("cache-control", "no-cache, no-transform")
				w.Header().Set("connection", "keep-alive")
				w.WriteHeader(http.StatusOK)
				headersSent = true
			}
			for _, out := range xlate.Feed(frame) {
				if _, err := w.Write(out); err != nil {
					return nil
				}
			}
			if f, ok := w.(http.Flusher); ok {
				f.Flush()
			}
		case err := <-errc:
			if !headersSent {
				_ = usage.Record(ctx, errEvent(ue, started, err))
				return err
			}
			logger.Warn("openai stream error after headers sent", "err", err)
			goto flush
		case <-ctx.Done():
			return nil
		}
	}
flush:
	for _, out := range xlate.Flush() {
		_, _ = w.Write(out)
	}
	if f, ok := w.(http.Flusher); ok {
		f.Flush()
	}
	prompt, completion := xlate.Usage()
	pricing := usage.GetPricing(ctx, ue.TenantID, ue.Provider, ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	return nil
}

// --- anthropic-compat passthrough (third-party endpoints mimicking /v1/messages) ---

func (d *Deps) nonStreamMessagesViaAnthropicCompat(ctx context.Context, w http.ResponseWriter, baseURL, apiKey string, metadata, body map[string]any, ue *usage.Event, started time.Time) error {
	result, err := d.AnthropicCompat.ChatMessages(ctx, baseURL, apiKey, metadata, body)
	if err != nil {
		_ = usage.Record(ctx, errEvent(ue, started, err))
		return err
	}
	u, _ := result["usage"].(map[string]any)
	prompt := toI64Or(u, "input_tokens")
	completion := toI64Or(u, "output_tokens")
	pricing := usage.GetPricing(ctx, ue.TenantID, ue.Provider, ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	httpx.WriteJSON(w, http.StatusOK, result)
	return nil
}

func (d *Deps) streamMessagesViaAnthropicCompat(ctx context.Context, w http.ResponseWriter, baseURL, apiKey string, metadata, body map[string]any, ue *usage.Event, started time.Time) error {
	frames, errc := d.AnthropicCompat.StreamMessages(ctx, baseURL, apiKey, metadata, body)
	var prompt, completion int64
	headersSent := false
	for {
		select {
		case frame, ok := <-frames:
			if !ok {
				goto done
			}
			if !headersSent {
				w.Header().Set("content-type", "text/event-stream")
				w.Header().Set("cache-control", "no-cache, no-transform")
				w.Header().Set("connection", "keep-alive")
				w.WriteHeader(http.StatusOK)
				headersSent = true
			}
			_, _ = w.Write(frame)
			if f, ok := w.(http.Flusher); ok {
				f.Flush()
			}
			p, c := scanAnthropicUsage(frame)
			if p > 0 {
				prompt = p
			}
			if c > 0 {
				completion = c
			}
		case err := <-errc:
			if !headersSent {
				_ = usage.Record(ctx, errEvent(ue, started, err))
				return err
			}
			logger.Warn("anthropic-compat stream error after headers sent", "err", err)
			goto done
		case <-ctx.Done():
			return nil
		}
	}
done:
	pricing := usage.GetPricing(ctx, ue.TenantID, ue.Provider, ue.UpstreamModel)
	ue.Status, ue.LatencyMs = "ok", time.Since(started).Milliseconds()
	ue.PromptTokens, ue.CompletionTokens = prompt, completion
	ue.CostMicros = usage.ComputeCost(pricing, prompt, completion)
	_ = usage.Record(ctx, ue)
	return nil
}

func toI64Or(m map[string]any, k string) int64 {
	if m == nil {
		return 0
	}
	return toI64(m[k])
}
